File: batch_processor/csv_writer.py
"""
CSV output generation for evaluation results.
"""

import csv
from pathlib import Path
from typing import List
from decimal import Decimal

from risk_core.models import PortfolioResult, Asset, Portfolio, Scenario
from batch_processor.logger import get_logger

logger = get_logger(__name__)


class CSVWriter:
    """Writes evaluation results to CSV files."""
    
    CSV_HEADERS = [
        'run_id',
        'timestamp',
        'portfolio_id',
        'scenario_id',
        'asset_id',
        'currency',
        'base_value',
        'adjusted_value',
        'value_change',
        'value_change_percent',
        'fx_impact',
        'market_impact',
        'volatility_impact'
    ]
    
    def write_results(self, results: List[PortfolioResult], output_file: Path) -> None:
        """
        Write portfolio results to CSV file.
        
        Args:
            results: List of PortfolioResult objects
            output_file: Path to output CSV file
        """
        logger.info(f"Writing {len(results)} result(s) to {output_file}")
        
        with open(output_file, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            
            # Write header
            writer.writerow(self.CSV_HEADERS)
            
            # Write rows for each result
            for result in results:
                self._write_result_rows(writer, result)
        
        logger.info(f"CSV file written successfully: {output_file}")
    
    def _write_result_rows(self, writer: csv.writer, result: PortfolioResult) -> None:
        """
        Write rows for a single portfolio result.
        
        Args:
            writer: CSV writer object
            result: PortfolioResult to write
        """
        # TODO: This needs access to the original portfolio and scenario
        # to write asset-level details. For now, write portfolio-level summary.
        # In full implementation, this would iterate through assets.
        
        # Write portfolio-level summary row
        row = [
            result.run_id,
            result.timestamp,
            result.portfolio_id or '',
            result.scenario_id or '',
            'PORTFOLIO_TOTAL',  # Asset ID placeholder
            '',  # Currency (multi-currency)
            str(result.base_portfolio_value),
            str(result.adjusted_portfolio_value),
            str(result.value_change),
            str(result.value_change_percent),
            '',  # FX impact (calculated per currency)
            '',  # Market impact (calculated per asset)
            ''   # Volatility impact (calculated per asset)
        ]
        writer.writerow(row)
        
        # TODO: Write individual asset rows
        # This requires storing asset-level results in PortfolioResult
        # or passing portfolio/scenario to this method
    
    def write_result_with_assets(
        self,
        result: PortfolioResult,
        portfolio: Portfolio,
        scenario: Scenario,
        asset_results: dict,
        output_file: Path
    ) -> None:
        """
        Write detailed results including asset-level breakdown.
        
        Args:
            result: PortfolioResult object
            portfolio: Original portfolio
            scenario: Applied scenario
            asset_results: Dictionary mapping asset IDs to adjusted values
            output_file: Path to output CSV file
        """
        with open(output_file, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(self.CSV_HEADERS)
            
            # Write asset-level rows
            for asset in portfolio.assets:
                adjusted_value = asset_results.get(asset.id, asset.value)
                value_change = adjusted_value - asset.value
                value_change_percent = (
                    (value_change / asset.value * Decimal('100'))
                    if asset.value != Decimal('0') else Decimal('0')
                )
                
                # Calculate FX impact
                fx_impact = Decimal('0')
                if asset.currency in scenario.currency_rate_changes:
                    rate_change_bps = scenario.currency_rate_changes[asset.currency]
                    fx_impact = asset.value * (rate_change_bps / Decimal('10000'))
                
                # Calculate market impact
                market_impact = asset.value * (scenario.market_movement_bps / Decimal('10000'))
                
                # Calculate volatility impact
                volatility_impact = asset.value * (scenario.volatility_change_bps / Decimal('10000'))
                
                row = [
                    result.run_id,
                    result.timestamp,
                    result.portfolio_id or '',
                    result.scenario_id or '',
                    asset.id,
                    asset.currency,
                    str(asset.value),
                    str(adjusted_value),
                    str(value_change),
                    str(value_change_percent),
                    str(fx_impact),
                    str(market_impact),
                    str(volatility_impact)
                ]
                writer.writerow(row)

File: batch_processor_cpp/src/csv_writer.cpp
#include "batch_processor/csv_writer.hpp"
#include "risk_core/currency.hpp"
#include "risk_core/scenario.hpp"
#include <iomanip>
#include <sstream>
#include <stdexcept>

namespace batch_processor {

const std::vector<std::string> CsvWriter::CSV_HEADERS = {
    "run_id", "timestamp", "portfolio_id", "scenario_id",
    "asset_id", "currency", "base_value", "adjusted_value",
    "value_change", "value_change_percent", "fx_impact",
    "market_impact", "volatility_impact"
};

void CsvWriter::write_results(
    const std::vector<risk_core::PortfolioResult>& results,
    const risk_core::Portfolio& portfolio,
    const std::vector<risk_core::Scenario>& scenarios,
    const std::string& output_file
) {
    std::ofstream file(output_file);
    if (!file.is_open()) {
        throw std::runtime_error("Failed to open output file: " + output_file);
    }
    
    write_header(file);
    
    // Create scenario map for quick lookup
    std::map<std::string, const risk_core::Scenario*> scenario_map;
    for (const auto& scenario : scenarios) {
        scenario_map[scenario.scenario_id()] = &scenario;
    }
    
    // Write asset-level rows
    for (const auto& result : results) {
        const risk_core::Scenario* scenario = nullptr;
        if (!result.scenario_id().empty()) {
            auto it = scenario_map.find(result.scenario_id());
            if (it != scenario_map.end()) {
                scenario = it->second;
            }
        }
        
        // Apply scenario to get adjusted values
        if (scenario) {
            risk_core::ScenarioApplicator applicator(*scenario);
            auto adjusted_values = applicator.apply_to_portfolio(portfolio, "USD");
            
            for (const auto& asset : portfolio.assets()) {
                double adjusted_value = adjusted_values.at(asset.id());
                write_asset_row(file, result, asset, adjusted_value, *scenario);
            }
        }
    }
    
    file.close();
}

void CsvWriter::write_header(std::ofstream& file) {
    for (size_t i = 0; i < CSV_HEADERS.size(); ++i) {
        if (i > 0) file << ",";
        file << CSV_HEADERS[i];
    }
    file << "\n";
}

void CsvWriter::write_asset_row(
    std::ofstream& file,
    const risk_core::PortfolioResult& result,
    const risk_core::Asset& asset,
    double adjusted_value,
    const risk_core::Scenario& scenario
) {
    double value_change = adjusted_value - asset.value();
    double value_change_percent = risk_core::calculate_value_change_percent(
        asset.value(), adjusted_value
    );
    
    double fx_impact = calculate_fx_impact(asset, scenario, "USD");
    double market_impact = calculate_market_impact(asset, scenario);
    double volatility_impact = calculate_volatility_impact(asset, scenario);
    
    file << result.run_id() << ","
         << result.timestamp() << ","
         << result.portfolio_id() << ","
         << result.scenario_id() << ","
         << asset.id() << ","
         << asset.currency() << ","
         << std::fixed << std::setprecision(2) << asset.value() << ","
         << adjusted_value << ","
         << value_change << ","
         << value_change_percent << ","
         << fx_impact << ","
         << market_impact << ","
         << volatility_impact << "\n";
}

double CsvWriter::calculate_fx_impact(
    const risk_core::Asset& asset,
    const risk_core::Scenario& scenario,
    const std::string& base_currency
) {
    if (asset.currency() == base_currency) {
        return 0.0;
    }
    
    double rate_change_bps = scenario.get_currency_rate_change(asset.currency());
    if (rate_change_bps != 0.0) {
        return asset.value() * (rate_change_bps / 10000.0);
    }
    return 0.0;
}

double CsvWriter::calculate_market_impact(
    const risk_core::Asset& asset,
    const risk_core::Scenario& scenario
) {
    return asset.value() * (scenario.market_movement_bps() / 10000.0);
}

double CsvWriter::calculate_volatility_impact(
    const risk_core::Asset& asset,
    const risk_core::Scenario& scenario
) {
    return asset.value() * (scenario.volatility_change_bps() / 10000.0);
}

} // namespace batch_processor

